terraform/README.md

```diff
@@ -21,12 +21,11 @@ The key features of Terraform are:
 
 For more information, see the [introduction section](http://www.terraform.io/intro) of the Terraform website.
 
-Getting Started & Documentation
--------------------------------
+ ## [`graft_net.sh`]
+Usage:  graft_net.sh provision |  reprovision | deploy | redeploy | destroy | status
```

Comment:
Please provide more detailed documentation on this. Questions are:
1. In general - what do we need to start using it (supposing we already have AWS account);
2. what exactly `provision` does, what input parameters needed (e.g. credentials, keys, source ami's, instance types, etc), what is expected as outcome ( e.g. number of running ec2 instances, with or without graft packages deployed)
3. what `reprovision` does and what is expected outcome? 
4. what  exactly `deploy` does, where it looks for source packages

Reply:
UPD: after setting up env variable with AWS credentials and invoking `./graft_net.sh status` :

```
mbg033@n550jk ~/.../terraform (fix/debian-packaging *)$] ./graft_net.sh status
jq has been found
[DEPRECATION WARNING]: ANSIBLE_HOSTS option, The variable is misleading as it can be a list of hosts and/or paths to inventory sources , use ANSIBLE_INVENTORY instead. This feature will be removed in version 2.8. Deprecation warnings
 can be disabled by setting deprecation_warnings=False in ansible.cfg.
ERROR! the playbook: service.yml could not be found
```

Reply:
1. Environment and Prerequisites description not found
2. Architecture description not found
3. Installation Guide not found
4. Making configuration for Deployment not found
5. There is no environment check, there are no messages that there is not enough environment, during installation, the system “crashes” without messages.

Reply:
> UPD: after setting up env variable with AWS credentials and invoking `./graft_net.sh status`

Steps are:
1. have ubuntu 18.04 x64;
2. install `ansible` with  `sudo apt install ansible`
3. install aws CLI tools as described here: https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html#install-tool-pip
4.  get the project files:
    ``` 
    git checkout -b angrysham-master master
    git pull https://github.com/angrysham/GraftTools.git master
    ```
5. export AWS credentials as 
   ```
   export AWS_ACCESS_KEY_ID=xxxxx
   export AWS_SECRET_ACCESS_KEY=xxxxxx
   ``` 
6. goto `terraform` directory, invoke `./graft_net.sh status`:
    ```
    [mbg033@n550jk ~/.../GraftTools (angrysham-master)$] cd terraform/
    [mbg033@n550jk ~/.../terraform (angrysham-master)$] ./graft_net.sh status
    jq has been found
    [DEPRECATION WARNING]: ANSIBLE_HOSTS option, The variable is misleading as it can be 
   a list of hosts and/or paths to inventory sources , use ANSIBLE_INVENTORY instead. This 
   feature will be removed in version 2.8. Deprecation warnings
   can be disabled by setting deprecation_warnings=False in ansible.cfg.
   ERROR! the playbook: service.yml could not be found
   [mbg033@n550jk ~/.../terraform (angrysham-master)$] 
   ```

Reply:
> 2) To generate and upload deb package one can use aws_build folder and supplementary scripts to run build and upload artifacts to predefined Nexus
Debian Repo

Couldn't find in pdf where to setup URL/credentials for APT repo

Reply:
Trying to run a `graft_net.sh provision`:

```console
[mbg033@n550jk ~/.../terraform (angrysham-master *)$] ./graft_net.sh provision
jq has been found
Initializing modules...

Initializing the backend...

Initializing provider plugins...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.aws: version = "~> 2.15"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

Error: Unsupported block type

  on ../modules/networking/main.tf line 6, in resource "aws_vpc" "vpc":
   6:   tags {

Blocks of type "tags" are not expected here. Did you mean to define argument
"tags"? If so, use the equals sign to assign it a value.
```

Reply:
From the documentation:
```
supernode_instance_count - Number of suppernode instances
cryptonode_instance_count - Number of graftnode instances
```
We don't need a different types - both packages (cryptonode, supernode) supposed to be deployed into one instance

Reply:
Also, once it will be successfully provisioned, not clear how end-user should access every deployed instance via ssh? will it be public IPs or something else?

Reply:
> 
> 
> > 1. To generate and upload deb package one can use aws_build folder and supplementary scripts to run build and upload artifacts to predefined Nexus
> >    Debian Repo
> 
> Couldn't find in pdf where to setup URL/credentials for APT repo
APT repo has been preinstalled during deployment phase; credentials have been shared to Boris

Reply:
> 
> 
> Trying to run a `graft_net.sh provision`:
> 
> ```
> [mbg033@n550jk ~/.../terraform (angrysham-master *)$] ./graft_net.sh provision
> jq has been found
> Initializing modules...
> 
> Initializing the backend...
> 
> Initializing provider plugins...
> 
> The following providers do not have any version constraints in configuration,
> so the latest version was installed.
> 
> To prevent automatic upgrades to new major versions that may contain breaking
> changes, it is recommended to add version = "..." constraints to the
> corresponding provider blocks in configuration, with the constraint strings
> suggested below.
> 
> * provider.aws: version = "~> 2.15"
> 
> Terraform has been successfully initialized!
> 
> You may now begin working with Terraform. Try running "terraform plan" to see
> any changes that are required for your infrastructure. All Terraform commands
> should now work.
> 
> If you ever set or change modules or backend configuration for Terraform,
> rerun this command to reinitialize your working directory. If you forget, other
> commands will detect it and remind you to do so if necessary.
> 
> Error: Unsupported block type
> 
>   on ../modules/networking/main.tf line 6, in resource "aws_vpc" "vpc":
>    6:   tags {
> 
> Blocks of type "tags" are not expected here. Did you mean to define argument
> "tags"? If so, use the equals sign to assign it a value.
> ```

This kind of error can be reproduced with terraform releases > 0.12.X
It works and creates infrastructure for 0.11.8 version( https://releases.hashicorp.com/terraform/0.11.8/ )

Reply:
> 
> 
> Also, once it will be successfully provisioned, not clear how end-user should access every deployed instance via ssh? will it be public IPs or something else?

Instances can be accessed via Bastion host only as it's gonna be deployed in a private VPC subnet. Ansible uses dynamic inventory file and ENV filter based on sandbox name to deploy debian packages and start appropriate services according to a tag assigned during provisioning phase(terraform).

Reply:
Thanks, after installing 0.11.8 `./graft-net.sh provision` completed; 

> Instances can be accessed via Bastion host only as it's gonna be deployed in a private VPC subnet. Ansible uses dynamic inventory file and ENV filter based on sandbox name to deploy debian packages and start appropriate services according to a tag assigned during provisioning phase(terraform).

Some points still not clear:
- are there documented steps for our developers needs to be  performed to connect to one of the running instance (e.g.: to check how cryptonode/supernode running)
- is there an option to run the instances on public IPs?
- is it possible to use this CI toolset for developer who doesn't have access to AWS console (via web-interface)
- `graft-net.sh deploy` - should this command install/update packages from "Preinstalled APT repo"? I run this and got following:
```

[mbg033@n550jk ~/.../terraform (angrysham-master *)$] ./graft_net.sh deploy
jq has been found
++ aws ec2 describe-instances --filters Name=tag:Name,Values=sandbox1-bastion --output=json
++ jq -r '.Reservations[].Instances[].NetworkInterfaces[].Association.PublicIp'
+ public_IP='54.173.100.64
52.90.107.162'
+ sed -e '/ProxyJump Proxy_var/c\  ProxyJump 54.173.100.64
52.90.107.162' -e 's/Host Proxy_var/Host 54.173.100.64
52.90.107.162/g' ssh_config_template
sed: -e expression #1, char 52: unknown command: `.'
Installing Python2...
[DEPRECATION WARNING]: ANSIBLE_HOSTS option, The variable is misleading as it can be a list of hosts and/or paths to inventory sources , use ANSIBLE_INVENTORY instead. This feature will be removed in version 2.8. Deprecation warnings
 can be disabled by setting deprecation_warnings=False in ansible.cfg.
 [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with script plugin: Inventory script (/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py) had an execution error:
Traceback (most recent call last):   File "/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py", line 162, in <module>     import boto ImportError: No module named boto

 [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with ini plugin: /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py:4: Expected key=value host variable assignment,
got: external

 [WARNING]: Unable to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available


PLAY [all] *******************************************************************************************************************************************************************************************************************************

TASK [Install python2 for Ansible] *******************************************************************************************************************************************************************************************************
fatal: []: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: ssh: Could not resolve hostname : Name or service not known\r\n", "unreachable": true}
        to retry, use: --limit @/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/common.retry

PLAY RECAP *******************************************************************************************************************************************************************************************************************************
                           : ok=0    changed=0    unreachable=1    failed=0   

Deploying graftnode package to Role: cryptonode hosts...
[DEPRECATION WARNING]: ANSIBLE_HOSTS option, The variable is misleading as it can be a list of hosts and/or paths to inventory sources , use ANSIBLE_INVENTORY instead. This feature will be removed in version 2.8. Deprecation warnings
 can be disabled by setting deprecation_warnings=False in ansible.cfg.
 [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with script plugin: Inventory script (/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py) had an execution error:
Traceback (most recent call last):   File "/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py", line 162, in <module>     import boto ImportError: No module named boto

 [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with ini plugin: /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py:4: Expected key=value host variable assignment,
got: external

 [WARNING]: Unable to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: Could not match supplied host pattern, ignoring: tag_Role_cryptonode


PLAY [tag_Role_cryptonode] ***************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY RECAP *******************************************************************************************************************************************************************************************************************************

Deploying graftnode package to Role: supernode hosts...
[DEPRECATION WARNING]: ANSIBLE_HOSTS option, The variable is misleading as it can be a list of hosts and/or paths to inventory sources , use ANSIBLE_INVENTORY instead. This feature will be removed in version 2.8. Deprecation warnings
 can be disabled by setting deprecation_warnings=False in ansible.cfg.
 [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with script plugin: Inventory script (/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py) had an execution error:
Traceback (most recent call last):   File "/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py", line 162, in <module>     import boto ImportError: No module named boto

 [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with ini plugin: /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py:4: Expected key=value host variable assignment,
got: external

 [WARNING]: Unable to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available


PLAY [all] *******************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *******************************************************************************************************************************************************************************************************************
fatal: []: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: ssh: Could not resolve hostname : Name or service not known\r\n", "unreachable": true}
        to retry, use: --limit @/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/deploy-supernode.retry

PLAY RECAP *******************************************************************************************************************************************************************************************************************************
                           : ok=0    changed=0    unreachable=1    failed=0   
```
it's about passed since "provision" finished.

Reply:
> 
> 
> Thanks, after installing 0.11.8 `./graft-net.sh provision` completed;
> 
> > Instances can be accessed via Bastion host only as it's gonna be deployed in a private VPC subnet. Ansible uses dynamic inventory file and ENV filter based on sandbox name to deploy debian packages and start appropriate services according to a tag assigned during provisioning phase(terraform).
> 
> Some points still not clear:
> 
>     * are there documented steps for our developers needs to be  performed to connect to one of the running instance (e.g.: to check how cryptonode/supernode running)
> 
>     * is there an option to run the instances on public IPs?
> 
>     * is it possible to use this CI toolset for developer who doesn't have access to AWS console (via web-interface)
> 
>     * `graft-net.sh deploy` - should this command install/update packages from "Preinstalled APT repo"? I run this and got following:
> 
> 
> ```
> 
> [mbg033@n550jk ~/.../terraform (angrysham-master *)$] ./graft_net.sh deploy
> jq has been found
> ++ aws ec2 describe-instances --filters Name=tag:Name,Values=sandbox1-bastion --output=json
> ++ jq -r '.Reservations[].Instances[].NetworkInterfaces[].Association.PublicIp'
> + public_IP='54.173.100.64
> 52.90.107.162'
> + sed -e '/ProxyJump Proxy_var/c\  ProxyJump 54.173.100.64
> 52.90.107.162' -e 's/Host Proxy_var/Host 54.173.100.64
> 52.90.107.162/g' ssh_config_template
> sed: -e expression #1, char 52: unknown command: `.'
> Installing Python2...
> [DEPRECATION WARNING]: ANSIBLE_HOSTS option, The variable is misleading as it can be a list of hosts and/or paths to inventory sources , use ANSIBLE_INVENTORY instead. This feature will be removed in version 2.8. Deprecation warnings
>  can be disabled by setting deprecation_warnings=False in ansible.cfg.
>  [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with script plugin: Inventory script (/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py) had an execution error:
> Traceback (most recent call last):   File "/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py", line 162, in <module>     import boto ImportError: No module named boto
> 
>  [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with ini plugin: /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py:4: Expected key=value host variable assignment,
> got: external
> 
>  [WARNING]: Unable to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py as an inventory source
> 
>  [WARNING]: No inventory was parsed, only implicit localhost is available
> 
> 
> PLAY [all] *******************************************************************************************************************************************************************************************************************************
> 
> TASK [Install python2 for Ansible] *******************************************************************************************************************************************************************************************************
> fatal: []: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: ssh: Could not resolve hostname : Name or service not known\r\n", "unreachable": true}
>         to retry, use: --limit @/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/common.retry
> 
> PLAY RECAP *******************************************************************************************************************************************************************************************************************************
>                            : ok=0    changed=0    unreachable=1    failed=0   
> 
> Deploying graftnode package to Role: cryptonode hosts...
> [DEPRECATION WARNING]: ANSIBLE_HOSTS option, The variable is misleading as it can be a list of hosts and/or paths to inventory sources , use ANSIBLE_INVENTORY instead. This feature will be removed in version 2.8. Deprecation warnings
>  can be disabled by setting deprecation_warnings=False in ansible.cfg.
>  [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with script plugin: Inventory script (/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py) had an execution error:
> Traceback (most recent call last):   File "/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py", line 162, in <module>     import boto ImportError: No module named boto
> 
>  [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with ini plugin: /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py:4: Expected key=value host variable assignment,
> got: external
> 
>  [WARNING]: Unable to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py as an inventory source
> 
>  [WARNING]: No inventory was parsed, only implicit localhost is available
> 
>  [WARNING]: Could not match supplied host pattern, ignoring: tag_Role_cryptonode
> 
> 
> PLAY [tag_Role_cryptonode] ***************************************************************************************************************************************************************************************************************
> skipping: no hosts matched
> 
> PLAY RECAP *******************************************************************************************************************************************************************************************************************************
> 
> Deploying graftnode package to Role: supernode hosts...
> [DEPRECATION WARNING]: ANSIBLE_HOSTS option, The variable is misleading as it can be a list of hosts and/or paths to inventory sources , use ANSIBLE_INVENTORY instead. This feature will be removed in version 2.8. Deprecation warnings
>  can be disabled by setting deprecation_warnings=False in ansible.cfg.
>  [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with script plugin: Inventory script (/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py) had an execution error:
> Traceback (most recent call last):   File "/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py", line 162, in <module>     import boto ImportError: No module named boto
> 
>  [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with ini plugin: /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py:4: Expected key=value host variable assignment,
> got: external
> 
>  [WARNING]: Unable to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py as an inventory source
> 
>  [WARNING]: No inventory was parsed, only implicit localhost is available
> 
> 
> PLAY [all] *******************************************************************************************************************************************************************************************************************************
> 
> TASK [Gathering Facts] *******************************************************************************************************************************************************************************************************************
> fatal: []: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: ssh: Could not resolve hostname : Name or service not known\r\n", "unreachable": true}
>         to retry, use: --limit @/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/deploy-supernode.retry
> 
> PLAY RECAP *******************************************************************************************************************************************************************************************************************************
>                            : ok=0    changed=0    unreachable=1    failed=0   
> ```
> 
> it's about passed since "provision" finished.

>are there documented steps for our developers needs to be performed to connect to one of the running instance (e.g.: to check how cryptonode/supernode running)
it's gonna be checked with status/redeploy command

> is there an option to run the instances on public IPs?
yes, but it requires changes to TF files. By design sandbox is deployed in isolated VPC subnet. To access it one can use Bastion host.

> is it possible to use this CI toolset for developer who doesn't have access to AWS console (via web-interface)
It's been planned as Phase 2 to be triggered from Jenkinks 

for deploy -- looks like dynamic provisioning process doesn't return hosts and tags
check output of  `./ec2.py --list` command from ansible folder

Reply:
> 
> it's about passed since "provision" finished.

> 52.90.107.162/g' ssh_config_template
> **sed: -e expression #1, char 52: unknown command: `.'**


> **fatal: []: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: ssh: Could not resolve hostname : Name or service not known\r\n", "unreachable": true}
>         to retry, use: --limit @/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/deploy-supernode.retry**

>**[WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with script plugin: Inventory script (/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py) had an execution error:
> Traceback (most recent call last):   File "/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py", line 162, in <module>     import boto ImportError: No module named boto**

looks like it's not completely passed?

Reply:
> > are there documented steps for our developers needs to be performed to connect to one of the running instance (e.g.: to check how cryptonode/supernode running)

> 
> it's gonna be checked with status/redeploy command
Not clear - literally developer need to ssh to the instance to check the logs - can we have a documented steps how to do this?

Also, looks like something wrong with "status" command?

```
[mbg033@n550jk ~/.../terraform (angrysham-master *)$] ./graft_net.sh status                                                                                                                                                               
jq has been found                                                                                                                                                                                                                         
[DEPRECATION WARNING]: ANSIBLE_HOSTS option, The variable is misleading as it can be a list of hosts and/or paths to inventory sources , use ANSIBLE_INVENTORY instead. This feature will be removed in version 2.8. Deprecation warnings 
 can be disabled by setting deprecation_warnings=False in ansible.cfg.                                                                                                                                                                    
 [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with script plugin: Inventory script (/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py) had an execution error:       
Traceback (most recent call last):   File "/home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py", line 162, in <module>     import boto ImportError: No module named boto                                                   
                                                                                                                                                                                                                                          
 [WARNING]:  * Failed to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py with ini plugin: /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py:4: Expected key=value host variable assignment,     
got: external                                                                                                                                                                                                                             
                                                                                                                                                                                                                                          
 [WARNING]: Unable to parse /home/mbg033/dev/graft-project/GraftTools/terraform/ansible/ec2.py as an inventory source                                                                                                                     
                                                                                                                                                                                                                                          
 [WARNING]: No inventory was parsed, only implicit localhost is available                                                                                                                                                                 
                                                                                                                                                                                                                                          
 [WARNING]: Could not match supplied host pattern, ignoring: tag_Role_cryptonode                                                                                                                                                          
                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                          
PLAY [tag_Role_cryptonode] ***************************************************************************************************************************************************************************************************************
skipping: no hosts matched                                                                                                                                                                                                                
                                                                                                                                                                                                                                          
PLAY RECAP *******************************************************************************************************************************************************************************************************************************
                                                                                                                                                                                                                                          
Press any key to continue...            
```

> for deploy -- looks like dynamic provisioning process doesn't return hosts and tags
> check output of `./ec2.py --list` command from ansible folder
```
[mbg033@n550jk ~/.../ansible (angrysham-master *)$] ./ec2.py --list
Traceback (most recent call last):                                                                                                                                                                                                        
  File "./ec2.py", line 162, in <module>                                                                                                                                                                                                  
    import boto                                                                                                                                                                                                                           
ImportError: No module named boto   
```
Looks like some missing steps in documentation, please fix it

Reply:
Resuming: 
- please update documentation with the steps how developer should ssh to the any of "sanbox" instances
- seems like "sandbox1" tag is hardcoded somewhere - is it possible so we run many such independent "sandboxes" - if yes - please update documentation with the steps how to change "sandbox1" to something else.

Reply:
> 
> 
> Resuming:
> 
>     * there're errors/warnings while invoking following commands:  `./graft_net.sh status`,  `./graft_net.sh deploy` and ` ./ec2.py --list` (see above)
>
please specify which python and aws-cli versions u're using for this step 

>     * please update documentation with the steps how developer should ssh to the any of "sanbox" instances;
it wasn't purpose to allow developer perform direct login, but using list of nodes from ec2.py and ssh_config file it's possible to connect to any node being deployed in sandbox VPC 
> 
>     * "sandbox1" - is it possible to change it so we can have many such environments simultaneously? if yes - please document the steps
yes, it possible by default one can create only 4 custom VPC per region == 4 sandbox environments. One can request limit change for a region to a needed number of sandboxes...

Reply:
> > Resuming:
> > ```
> > * there're errors/warnings while invoking following commands:  `./graft_net.sh status`,  `./graft_net.sh deploy` and ` ./ec2.py --list` (see above)
> > ```
> 
> please specify which python and aws-cli versions u're using for this step

python: having both 2 and 3 version installed, default is version 2
```
[mbg033@n550jk ~$] ls -la $(which python)
lrwxrwxrwx 1 root root 9 окт 14  2018 /usr/bin/python -> python2.7
[mbg033@n550jk ~$] python --version
Python 2.7.15+
[mbg033@n550jk ~$] la -la $(which python3)
lrwxrwxrwx 1 root root 9 окт 25  2018 /usr/bin/python3 -> python3.6
[mbg033@n550jk ~$] python3 --version
Python 3.6.8
```

AWS CLI - installed using this [link](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html)
```
[mbg033@n550jk ~$] aws --version
aws-cli/1.16.173 Python/3.6.8 Linux/4.15.0-52-generic botocore/1.12.163
```

> > ```
> > * please update documentation with the steps how developer should ssh to the any of "sanbox" instances;
> > ```
> 
> it wasn't purpose to allow developer perform direct login, but using list of nodes from ec2.py and ssh_config file it's possible to connect to any node being deployed in sandbox VPC

Question is not about direct login - we need a documented steps so developer can ssh to remote instance running cryptonode/supernode. Doesn't matter directly or indirectly. Same way, developer needs to be able to perform to call some RPC/REST interfaces on cryptonode/supernode

 
> > ```
> > * "sandbox1" - is it possible to change it so we can have many such environments simultaneously? if yes - please document the steps
> > ```
> 
> yes, it possible by default one can create only 4 custom VPC per region == 4 sandbox environments. One can request limit change for a region to a needed number of sandboxes...

Can we have documented what needs to be changed so "sandbox1" (guess this is a tag or name on aws) will became e.g. "sandbox2" ? What steps developer need to perform to run 2nd and 3rd environment (within four VPC limit) from his machine?

Reply:
one can change variable in file 'sandbox1/variables.tf':
default = "sandbox1" 
to sandbox2, sandbox3 etc

Reply:
To perform login exec:
ssh -F ssh_config ubuntu@10.0.1.174
ssh_config - generates automatically from template by running 'graft_net.sh deploy' or mannually -- update_ssh_config.sh
IP addresses are obtained from 'ec2.py --list'